File: java/dnd-character/src/main/java/DnDCharacter.java
import java.util.Comparator;
import java.util.List;
import java.util.Random;
import java.util.stream.Collectors;
import java.util.stream.IntStream;

class DnDCharacter {

    private static final Random random = new Random();
    private final int strength;
    private final int dexterity;
    private final int constitution;
    private final int intelligence;
    private final int wisdom;
    private final int charisma;

    public DnDCharacter() {
        this.strength = ability(rollDice());
        this.dexterity = ability(rollDice());
        this.constitution = ability(rollDice());
        this.intelligence = ability(rollDice());
        this.wisdom = ability(rollDice());
        this.charisma = ability(rollDice());
    }

    int ability(List<Integer> scores) {
        return scores.stream().sorted(Comparator.reverseOrder()).limit(3).reduce(0, Integer::sum);
    }

    List<Integer> rollDice() {
        return IntStream.range(0, 4).mapToObj(x -> random.nextInt(6) + 1).limit(4).collect(Collectors.toList());
    }

    int modifier(int input) {
        return (int) Math.floor((double) (input - 10) / 2);
    }

    int getStrength() {
        return this.strength;
    }

    int getDexterity() {
        return this.dexterity;
    }

    int getConstitution() {
        return this.constitution;
    }

    int getIntelligence() {
        return this.intelligence;
    }

    int getWisdom() {
        return this.wisdom;
    }

    int getCharisma() {
        return this.charisma;
    }

    int getHitpoints() {
        return 10 + modifier(getConstitution());
    }
}

File: ruby/dnd-character/dnd_character.rb
class DndCharacter
  BASE_HITPOINTS = 10

  def self.modifier(stat)
    (stat - 10) / 2
  end

  def self.roll_d6
    rand(1..6)
  end

  def self.roll_attribute
    rolls = []
    4.times { rolls << roll_d6 }
    rolls.sort![1..].sum
  end

  private

  def initialize
    %i[strength dexterity constitution intelligence wisdom charisma].each do |attribute|
      instance_variable_set("@#{attribute}", self.class.roll_attribute)
    end

    @hitpoints = self.class.modifier(constitution) + BASE_HITPOINTS
  end

  public

  attr_reader :strength, :dexterity, :constitution, :intelligence, :wisdom, :charisma, :hitpoints
end
